src/domain/shared/Log.ts

```diff
@@ -0,0 +1,245 @@
+                return func(`[${messageType}]`, message);
+            }
+            return func(message);
+        };
+    }
+
+    // it is necessary to capture the console object here, for things like jest's mocks to work,
+    // hence the no-op looking lambda wrappers
+    static #_functions = new Map<LogLevel, LogFunction>([
+        /* eslint-disable @typescript-eslint/no-explicit-any */
+        [LogLevel.DEFAULT, ConsoleLoggerContext.#_typeFirst((...params: any[]) => console.log(...params))],
+        [LogLevel.DEBUG, ConsoleLoggerContext.#_typeFirst((...params: any[]) => console.debug(...params))],
+        [LogLevel.INFO, ConsoleLoggerContext.#_typeFirst((...params: any[]) => console.info(...params))],
+        [LogLevel.WARNING, ConsoleLoggerContext.#_typeFirst((...params: any[]) => console.warn(...params))],
+        [LogLevel.ERROR, ConsoleLoggerContext.#_typeFirst((...params: any[]) => console.error(...params))]
+        /* eslint-enable @typescript-eslint/no-explicit-any */
+    ]);
+
+    /* eslint-disable class-methods-use-this */
+    getFunction(level: LogLevel): LogFunction | undefined {
+        return ConsoleLoggerContext.#_functions.get(level);
+    }
+    /* eslint-enable class-methods-use-this */
+}
+
+/*@devdoc
+ *  The <code>StringLoggerContext</code> is a configuration for logging to a string.
+ *
+ *  @class Logger
+ */
+export class StringLoggerContext implements LoggerContext {
+    buffer = "";
+    #_functions = new Map<LogLevel, LogFunction>();
+
+    constructor() {
+        for (const level of allLogLevels) {
+            this.#_functions.set(level, (message: string, messageType?: string) => {
+                if (messageType) {
+                    this.buffer += `[${messageType}]`;
+                }
+                this.buffer += `[${LogLevel[level] as string}] ${message}\n`;
+            });
+        }
+    }
+
+    getFunction(level: LogLevel): LogFunction | undefined {
+        return this.#_functions.get(level);
+    }
+
+}
+
+/*@devdoc
+ *  The <code>Logger</code> class serves as a convenience utility and a centralized configuration point for logging.
+ *
+ *  @class Logger
+ */
+export class Logger {
+    #_context: LoggerContext;
+    #_activeFunctions = new Map<LogLevel, LogFunction>();
+    #_typeFilter?: Array<string | undefined>;
+
+    #_messageIds = new Map<string, unknown>();
+    #_typedMessageIds = new Map<string, unknown>();
+    #_messageFlags = new Map<unknown, boolean>();
+
+    /*@devdoc
+     *  Creates a logger instance with a specified context.
+     *  @param {context} LoggerContext - The context this instance will use for output,
+     *  which can also store any addition state, like an output buffer.
+     */
+    constructor(context: LoggerContext) {
+        this.#_context = context;
+        this.filterLevels(() => true);
+    }
+
+    /*@devdoc
+     *  Logs a message at an appropriate level, with optional type.
+     *  @param {level} LogLevel - the level to log at
```

Comment:
```suggestion
     *  @param {LogLevel} level - The level to log at.
```
The type should be inside the `{}`. And similarly elsewhere.